### ISSUE TITLE
TOB-K8S-019: kubelet can cause an Out of Memory error with a malicious manifest

### ISSUE THREAD

**Issue by cji** [kubernetes/kubernetes#81122]
This issue was reported in the [Kubernetes Security Audit Report](https://github.com/kubernetes/community/blob/master/wg-security-audit/findings/Kubernetes%20Final%20Report.pdf)

**Description**
When kubelet attempts to load a Pod manifest, it will read the manifest entirely into memory in an attempt to parse and validate the manifest. Due to this behavior, if a large manifest is provided to kubelet it will cause an Out-Of-Memory (OOM) error. This could cause other services on the system upon which kubelet is running to fail outside of the scope of Kubernetes.

```
func (s *sourceFile) extractFromDir(name string) ([]*v1.Pod, error) {
	dirents, err := filepath.Glob(filepath.Join(name, "[^.]*"))
	...
	sort.Strings(dirents)
	for _, path := range dirents {
		statInfo, err := os.Stat(path)
		...
		switch {
		...
		case statInfo.Mode().IsRegular():
			pod, err := s.extractFromFile(path)
			...
		...
		}
	}
	return pods, nil
}
```
Figure 14.1: The extractFromDir function which finds all manifest files within a manifest directory and extracts Pods using extractFromFile.

```
func (s *sourceFile) extractFromFile(filename string) (pod *v1.Pod, err error) {
	...
	file, err := os.Open(filename)
             ...
	defer file.Close()

	data, err := ioutil.ReadAll(file)
	if err != nil {
		return pod, err
	}
	...
}
```
Figure 14.2: The extractFromFile function, which attempts to read all of the manifest files.

Additionally, due to kubelet causing a system OOM, the kubelet process will be killed and restarted. This appears to prevent kubelet from realizing that it has failed to start a Pod despite the backoff period, leading to an infinite loop of OOM, ultimately rendering the system unresponsive.

**Exploit Scenario**
Alice configures kubelet to pull Pods from a manifest directory. Eve identifies Alice’s kubelet manifest directory and has sufficient permissions to place a file in the manifest directory. Eve places a malicious manifest file within the kubelet manifest directory. kubelet’s restart policy and process-ephemeral backoff period causes the machine to lock in an OOM loop.

**Recommendation**
Avoid loading arbitrary data into memory regardless of size. Limit the size of a valid manifest or inform the user when it consumes a substantial amount of memory, especially for manifests that are fetched from remote endpoints. Consider persisting backoff periods for each Pod to allow for consistency between restarts of kubelet.

**Anything else we need to know?**:

See #81146 for current status of all issues created from these findings.

The vendor gave this issue an ID of TOB-K8S-019 and it was finding 14 of the report.

The vendor considers this issue Medium Severity.

To view the original finding, begin on page 46 of the [Kubernetes Security Review Report](https://github.com/kubernetes/community/blob/master/wg-security-audit/findings/Kubernetes%20Final%20Report.pdf)

**Environment**:

- Kubernetes version: 1.13.4

**Comment by mattjmcnaughton** [comment]
/assign

happy to tackle this :)

**Comment by mattjmcnaughton** [comment]
Actually, this was fixed in https://github.com/kubernetes/kubernetes/commit/1f270ef4e29fe699300a67eb096b80590dfc877e#diff-950f824b35e7510183b4a0c23de66065.

@liggitt is the plan with these security findings to cherry-pick them back to earlier releases?
If not, I think we can consider this one addressed :) If it is, happy to create the cherry-picks.

Great work @haiyanmeng.

**Comment by liggitt** [comment]
Backports depend on the severity of the issue. In this case, an attacker with write access to the static pod manifest directory already effectively has root access to the machine (they can place a privileged pod manifest the kubelet will run). Limiting the size of the read manifest to avoid OOM is nice to have, but not critical.

/close

Will be fixed in v1.16.0 by #76518

**Comment by mattjmcnaughton** [comment]
On Thu, Aug 08, 2019 at 12:41:39PM -0700, Jordan Liggitt wrote:
> Backports depend on the severity of the issue. In this case, an attacker with write access to the static pod manifest directory already effectively has root access to the machine (they can place a privileged pod manifest the kubelet will run). Limiting the size of the read manifest to avoid OOM is nice to have, but not critical.
> 
> /close
> 
> Will be fixed in v1.16.0 by #76518
> 
> -- 
> You are receiving this because you were assigned.
> Reply to this email directly or view it on GitHub:
> https://github.com/kubernetes/kubernetes/issues/81122#issuecomment-519659796

Makes sense to me - thanks for closing :)
